# Diabetes Prediction
### Using Support Vector Machine

#### Workflow:

- We have diabetes data
- Data Pre-processing
- Train-Test Data Split
- SVM Classifier


Importing the libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score, classification_report

### Data Collection and Analysis

PIMA Diabetes Dataset

In [2]:
# Loading the dataset to a pandas DataFrame
diabetes_dataset = pd.read_csv('datasets/diabetesPIMA.csv')

In [3]:
diabetes_dataset.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
0            6      148             72             35        0  33.6  \
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [4]:
diabetes_dataset.shape

(768, 9)

This means there are 9 columns (8 features, 1 target) and 768 rows (examples)

In [5]:
# getting the statistical data
diabetes_dataset.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin   
count   768.000000  768.000000     768.000000     768.000000  768.000000  \
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [6]:
diabetes_dataset['Outcome'].value_counts()

Outcome
0    500
1    268
Name: count, dtype: int64

So, there are 268 people with diabetes (=1) and 500 non-diabetic people (=0)

In [7]:
diabetes_dataset.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin   
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000  \
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

Now, we will separate the data and labels

In [8]:
X = diabetes_dataset.drop(columns='Outcome', axis=1)
y = diabetes_dataset['Outcome']

In [9]:
y.head()

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

In [10]:
X.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI   
0            6      148             72             35        0  33.6  \
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33

In [11]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
dtypes: float64(2), int64(6)
memory usage: 48.1 KB


### Data Standardization

We minimize the differences in ranges of different features.

In [12]:
scaler = StandardScaler()

In [13]:
standardized_X = scaler.fit_transform(X)

In [14]:
print(standardized_X)

[[ 0.63994726  0.84832379  0.14964075 ...  0.20401277  0.46849198
   1.4259954 ]
 [-0.84488505 -1.12339636 -0.16054575 ... -0.68442195 -0.36506078
  -0.19067191]
 [ 1.23388019  1.94372388 -0.26394125 ... -1.10325546  0.60439732
  -0.10558415]
 ...
 [ 0.3429808   0.00330087  0.14964075 ... -0.73518964 -0.68519336
  -0.27575966]
 [-0.84488505  0.1597866  -0.47073225 ... -0.24020459 -0.37110101
   1.17073215]
 [-0.84488505 -0.8730192   0.04624525 ... -0.20212881 -0.47378505
  -0.87137393]]


Now that all data is in similar range, we can split the data into training data and testing data

In [15]:
X = standardized_X

In [16]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2 ,stratify=y, random_state=2, )

In [17]:
X_train.shape, y_train.shape

((614, 8), (614,))

In [18]:
X_test.shape, y_test.shape

((154, 8), (154,))

### Training the Model

In [19]:
classifier = svm.SVC(kernel='linear')

In [20]:
# Train the classifier (SVM)
classifier.fit(X_train, y_train)

SVC(kernel='linear')

### Model Evaluation

Accuracy Score

In [21]:
X_train_preds = classifier.predict(X_train)
X_test_preds = classifier.predict(X_test)

In [22]:
testing_acc = accuracy_score(y_test, X_test_preds)

print(f'Testing Score of Classifier : {testing_acc}')

Testing Score of Classifier : 0.7727272727272727


In [23]:
training_acc = accuracy_score(y_train, X_train_preds)

print(f'Training Score of Classifier : {training_acc}')

Training Score of Classifier : 0.7866449511400652


Our model scores 78.66% on training data and 77.27% on testing data

### Making a Predictive System

In [24]:
input_data = [4, 210, 92, 0, 0, 37.6, 0.191, 30]
# 4, 110, 92, 0, 0, 37.6, 0.191, 30 will be format of data input

# Change the input data to numpy array
input_data_array = np.asarray(input_data)

# reshape the array
reshaped_input_data_array = input_data_array.reshape(1, -1)

# Standardize input data
std_data = scaler.transform(reshaped_input_data_array)

print(f'Standardized Data : {std_data}')

prediction = classifier.predict(std_data)
print(f'Prediction : {prediction}')

if (prediction[0] == 1):
    print('Patient has high chances of having Diabetes')
else:
    print('Patient is not diabetic')

Standardized Data : [[ 0.04601433  2.7887468   1.18359575 -1.28821221 -0.69289057  0.71168975
  -0.84827977 -0.27575966]]
Prediction : [1]
Patient has high chances of having Diabetes


C:\Users\Kunjesh\Downloads\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


## Saving the Model
We will use `pickle` library to save the model and will deploy the model using Azure

In [25]:
import pickle

In [26]:
filename = 'diabetes_model.pkl'
pickle.dump(classifier, open(filename, 'wb'))

In [27]:
# Loading the saved model
loaded_model = pickle.load(open(filename, 'rb'))

In [28]:
input_data = [4, 210, 92, 0, 0, 37.6, 0.191, 30]
# 4, 110, 92, 0, 0, 37.6, 0.191, 30 will be format of data input

# Change the input data to numpy array
input_data_array = np.asarray(input_data)

# reshape the array
reshaped_input_data_array = input_data_array.reshape(1, -1)

# Standardize input data
std_data = scaler.transform(reshaped_input_data_array)

print(f'Standardized Data : {std_data}')

prediction = loaded_model.predict(std_data)
print(f'Prediction : {prediction}')

if (prediction[0] == 1):
    print('Patient has high chances of having Diabetes')
else:
    print('Patient is not diabetic')

Standardized Data : [[ 0.04601433  2.7887468   1.18359575 -1.28821221 -0.69289057  0.71168975
  -0.84827977 -0.27575966]]
Prediction : [1]
Patient has high chances of having Diabetes


C:\Users\Kunjesh\Downloads\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [33]:
X = diabetes_dataset.drop(columns='Outcome', axis=1)
for col in X.columns:
    print(col)

Pregnancies
Glucose
BloodPressure
SkinThickness
Insulin
BMI
DiabetesPedigreeFunction
Age
